[compiler-rt][hwasan] Add fiber switch for HwASan

Tomahawkd opened the pull request:
Currently HwASan has no fiber switch interface for coroutines. This PR adds fiber switch interfaces similar to ASan which helps to pass sp check correctly on unwinding.

The only difference is HwASan does not need a fake stack since tags can do the same thing (e.g., detect UAR). Interfaces are made identical with ASan's.

Also adds unit test which is similar to ASan with minor adjustments:

1. change `__asan_handle_no_return` to `__hwasan_handle_vfork`
2. remove huge stack test since `__hwasan_handle_vfork` has no stack size limitation.
3. use uninstrumented globals to simulate allocation since hwasan do not support tagged pointer while using `longjmp`

The testcase is tested on both x86 with alias mode enabled and aarch64.

fmayer commented:
Thank you for your patch.

As far as I can tell from https://patch-diff.githubusercontent.com/raw/llvm/llvm-project/pull/153822.patch, you are using an anonymous email address alias from GitHub. The LLVM Project requires contributors to submit patches with an email address that they are monitoring. See https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it/74223.

fmayer commented on compiler-rt/test/hwasan/TestCases/Linux/swapcontext_annotation.cpp:
```diff
@@ -0,0 +1,202 @@
+// RUN: %clangxx_hwasan -std=c++11 -lpthread -O3 %s -o %t && %run %t 2>&1 | FileCheck %t.checks --check-prefix LOOPCHECK
+
+//
+// This test is too subtle to try on non-x86 arch for now.
```
How useful is this though? HWASan is mostly used on AArch64.

Tomahawkd committed:
[compiler-rt][hwasan] Add fiber switch for HwASan

Tomahawkd committed:
Minor adjustment from reviewer:
1. remove named unused param in `__sanitizer_start_switch_fiber` and `__sanitizer_finish_switch_fiber`
2. code style in `__sanitizer_start_switch_fiber` and `__sanitizer_finish_switch_fiber`
3. fix memory order action in `Thread::FinishSwitchFiber`
4. use `__builtin_frame_address` instead of `char local`

Tomahawkd committed:
refine testcase to test on aarch64

Tomahawkd commented on compiler-rt/lib/hwasan/hwasan_thread.cpp:
```diff
@@ -119,6 +119,60 @@ void Thread::Destroy() {
+}
+
+void Thread::FinishSwitchFiber(uptr *bottom_old, uptr *size_old) {
+  if (!atomic_load(&stack_switching_, memory_order_relaxed)) {
```
Fixed both `Thread::StartSwitchFiber` and `Thread::FinishSwitchFiber`

Tomahawkd commented on compiler-rt/lib/hwasan/hwasan_thread.cpp:
```diff
@@ -119,6 +119,60 @@ void Thread::Destroy() {
+    return {stack_bottom_, stack_top_};
+  }
+  char local;
+  const uptr cur_stack = (uptr)&local;
```
Fixed

Tomahawkd commented on compiler-rt/lib/hwasan/hwasan_thread.cpp:
```diff
@@ -226,3 +280,34 @@ void PrintThreads() {
+  }
+  t->FinishSwitchFiber((uptr *)bottom_old, size_old);
+}
+}
```
Fixed

Tomahawkd commented on compiler-rt/lib/hwasan/hwasan_thread.cpp:
```diff
@@ -226,3 +280,34 @@ void PrintThreads() {
+void __sanitizer_start_switch_fiber(void **unused, const void *bottom,
+                                    uptr size) {
+  // this is just a placeholder which make the interface same as ASan
+  (void)unused;
```
Fixed

Tomahawkd commented on compiler-rt/lib/hwasan/hwasan_thread.cpp:
```diff
@@ -226,3 +280,34 @@ void PrintThreads() {
+void __sanitizer_finish_switch_fiber(void *unused, const void **bottom_old,
+                                     uptr *size_old) {
+  // this is just a placeholder which make the interface same as ASan
+  (void)unused;
```
Fixed

Tomahawkd commented on compiler-rt/lib/hwasan/hwasan_thread.cpp:
```diff
@@ -226,3 +280,34 @@ void PrintThreads() {
+                                    uptr size) {
+  // this is just a placeholder which make the interface same as ASan
+  (void)unused;
+  auto *t = GetCurrentThread();
```
Fixed

Tomahawkd commented on compiler-rt/test/hwasan/TestCases/Linux/swapcontext_annotation.cpp:
```diff
@@ -0,0 +1,202 @@
+void handler(int sig) { CallNoReturn(); }
+
+int main(int argc, char **argv) {
+  // removed huge stack test since hwasan has no huge stack limitations
```
Comment added in testcase

Tomahawkd committed:
clang format fix

Tomahawkd committed:
use uninstrumented global to simulate allocation

fmayer commented:
Please fix  formatting of test case

Tomahawkd committed:
clang format fix

Tomahawkd commented:
> > ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> > You can test this locally with the following command:
> > View the diff from clang-format here.
> 
> Please fix formatting

Fixed, however, my clang-format still complains about the `RUN` comments...

fmayer commented:
> > > ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> > > You can test this locally with the following command:
> > > View the diff from clang-format here.
> > 
> > 
> > Please fix formatting
> 
> Fixed, however, my clang-format still complains about the `RUN` comments...

Did you run `git clang-format HEAD^`?

Tomahawkd commented:
> > > > ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> > > > You can test this locally with the following command:
> > > > View the diff from clang-format here.
> > > 
> > > 
> > > Please fix formatting
> > 
> > 
> > Fixed, however, my clang-format still complains about the `RUN` comments...
> 
> Did you run `git clang-format HEAD^`?

Yeah, I ran this:

```
git-clang-format --style LLVM --diff HEAD~6 HEAD --extensions h,cpp -- compiler-rt/test/hwasan/TestCases/Linux/swapcontext_annotation.cpp compiler-rt/lib/hwasan/hwasan_interface_internal.h compiler-rt/lib/hwasan/hwasan_thread.cpp compiler-rt/lib/hwasan/hwasan_thread.h
```

The clang-format is built along with the project and the `git-clang-format` is from the project. Quite confused.

fmayer commented on compiler-rt/lib/hwasan/hwasan_thread.cpp:
```diff
@@ -119,6 +119,59 @@ void Thread::Destroy() {
+    *size_old = stack_top_ - stack_bottom_;
+  stack_bottom_ = next_stack_bottom_;
+  stack_top_ = next_stack_top_;
+  atomic_store(&stack_switching_, 0, memory_order_release);
```
not sure why this is before the assignments below. but it is the same in asan, so let's leave it like this for now.
